edx_solutions_api_integration/courses/views.py

```diff
@@ -2381,33 +2381,22 @@ def get(self, request, course_id):  # pylint: disable=W0613
         if not course_exists(course_id):
             raise Http404
 
-        response_data = []
+        user_id = self.request.query_params.get('user_id', None)
+        role = self.request.query_params.get('role', None)
         course_key = get_course_key(course_id)
         instructors = CourseInstructorRole(course_key).users_with_role()
-        for instructor in instructors:
-            response_data.append({'id': instructor.id, 'role': 'instructor'})
-
         staff = CourseStaffRole(course_key).users_with_role()
-        for admin in staff:
-            response_data.append({'id': admin.id, 'role': 'staff'})
-
         observers = CourseObserverRole(course_key).users_with_role()
-        for observer in observers:
-            response_data.append({'id': observer.id, 'role': 'observer'})
-
         assistants = CourseAssistantRole(course_key).users_with_role()
-        for assistant in assistants:
-            response_data.append({'id': assistant.id, 'role': 'assistant'})
-
-        user_id = self.request.query_params.get('user_id', None)
+        all_users = (instructors | staff | observers | assistants).annotate(role=F("courseaccessrole__role"))
         if user_id:
-            response_data = list([item for item in response_data if int(item['id']) == int(user_id)])
-
-        role = self.request.query_params.get('role', None)
+            all_users &= User.objects.filter(id=int(user_id))
         if role:
-            response_data = list([item for item in response_data if item['role'] == role])
-
-        return Response(response_data, status=status.HTTP_200_OK)
+            all_users &= User.objects.filter(courseaccessrole__role=role)
+        return Response(
+            all_users.values("id", "role"),
+            status=status.HTTP_200_OK
+        )
```

Comment:
I think we can probably call `.distinct` with the right fields and it'll fix. Will confirm later.

Reply:
FIltering by annotation and calling distinct worked for me.

Reply:
And thanks for catching that! Although it's possible the downstream client code would've handled it fine, it's definitely not expected.

Reply:
> FIltering by annotation and calling distinct worked for me.

Good to know!

Reply:
> And thanks for catching that! Although it's possible the downstream client code would've handled it fine, it's definitely not expected.
The first case with duplications is probably something that might be OK, but the second is definitely likely to cause errors.

Reply:
So I think the duplication and wrong results were happening because we were joining the role table twice, as the SQL below shows. I've removed the extra join and also call distinct just in case.

Reply:
Here is what's working for me:
```python
        all_users = (instructors | staff | observers | assistants).annotate(role=F("courseaccessrole__role"))
        if user_id:
            all_users = all_users.filter(id=int(user_id))
        if role:
            all_users = all_users.filter(role=role)
        return Response(
            all_users.distinct().values("id", "role"),
            status=status.HTTP_200_OK
        )
```

Reply:
Yeah thats the same as what Ive got, havent pushed it yet because my internet died. Am on the phone, will push later.